### PR TITLE
Add developer docs on embeds

### PR DESCRIPTION
**Pull request by initvector**
This update adds docs for the basics of creating a new site embed.

Closes vanilla/vanilla#7215

### CODE REVIEW

**Comment by initvector** [comment]
Hold on merging until the JavaScript implementation is confirmed and released.

File: content/developer/embeds/index.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+```javascript
+import { registerEmbed, IEmbedData, FOCUS_CLASS } from "@dashboard/embeds";
+
+// Setup image embeds.
+registerEmbed("image", imageRenderer);
+
+/**
+ * Render an image embed in the editor.
+ */
+export async function imageRenderer(element: HTMLElement, data: IEmbedData) {
```
Comment (charrondev) [inline content/developer/embeds/index.md]: This example is of an async function that doesn't actually do anything asynchronously. I think we should offer a plain synchronous example like this, as well as one that does stuff asynchronously.

File: content/developer/embeds/index.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+After the embed class has been registered, subsequent calls to the embed manager will use the class for URL scraping and markup generation. To test, you can use Vanilla's API v2 media endpoint to scrape a URL that should be handled by the new component.
+
+## Client-Side Components
+
+Client-side rendering is performed with JavaScript. The implementation of a client-side embed renderer is straightforward, primarily powered by a single renderer function.
+
+Vanilla's `registerEmbed` function is required to register the new renderer. It must be imported from `@dashboard/embeds`. In addition, the `IEmbedData` interface is also required for  constructing a proper render function signature. To import these two requirements, you may include the following:
+
+```javascript
```
Comment (charrondev) [inline content/developer/embeds/index.md]: These should all be changed to `typescript` or `ts`. The type definitions in the function signatures and interfaces are not part of javascript. I think we need to make that pretty

File: content/developer/embeds/index.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+After the embed class has been registered, subsequent calls to the embed manager will use the class for URL scraping and markup generation. To test, you can use Vanilla's API v2 media endpoint to scrape a URL that should be handled by the new component.
+
+## Client-Side Components
+
+Client-side rendering is performed with JavaScript. The implementation of a client-side embed renderer is straightforward, primarily powered by a single renderer function.
+
+### Creating a New Renderer
+
+Any new client-side rendering functions must accept two parameters: `element` and `data`. The `element` parameter is the HTML element that will contain the rendered embed, while `data` is expected to be an instance of `IEmbedData` (which must be imported from `@dashboard/embeds`), containing the data returned from a server-side embed component's `matchUrl` method. Renderer functions do not return a value. They are expected to render any new HTML elements by acting on the value of `element`.
```
Comment (charrondev) [inline content/developer/embeds/index.md]: Just to clarify IEmbedData is an interface not a class, so it's not really an instance. The following would be more accurate

> while `data` is expected to be an object matching the shape of the interface `IEmbedData` (this interface may imported from `@dashboard/embeds`). Data returned from the `POST /api/v2/media/scrape` endpoint (defined in the `Embed->matchUrl()` above) will already conform to the shape of `IEmbedData`.

File: content/developer/embeds/index.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+Renderers may be defined as asynchronous functions. This can be particularly beneficial when ensuring an additional script is available as part of the rendering process. For example, a Twitter embed renderer might want to include the official Twitter widget library and wait for it to load, before proceeding with rendering.
```
Comment (charrondev) [inline content/developer/embeds/index.md]: I'd add the following here:

> The `ensureScript()` function may be imported from `@dashboard/dom`. This method will resolve  almost immediately if the script is already present, and ensure that the script only gets loaded a single time, no matter how many times you call it.

File: content/developer/embeds/index.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+            renderMyEmbed(element, renderData);
+        });
+    }
+}
+```
+
+### Example
+
+All the pieces can be seen coming together in the image embed renderer.
```
Comment (charrondev) [inline content/developer/embeds/index.md]: I think we should have this example build on the generic one you started above.

**/plugins/my-plugin/src/scripts/my-embed.ts**
```typescript
import { registerEmbed, IEmbedData } from "@dashboard/embeds";
 
// Register the embed renderer for javascript contexts
registerEmbed("my-embed", renderMyEmbed);

// Handle rendering for html from server-side contexts.
convertExistingEmbeds();

// Convert existing "MyEmbed" embeds.
function convertExistingEmbeds() {
    // Grab all the unconverted embeds. The my-unconverted-embeds class was part of the MyEmbeds::renderData output.
    const unconverted = Array.from(document.querySelectorAll(".my-unconverted-embeds"));
    if (unconverted.length > 0) {
        unconverted.forEach(element => {
            // Get embed data out of the data attributes.
            const url = element.getAttribute("data-myembed") || "";
 
            const renderData: IEmbedData = {
                type: "my-embed",
                url,
            };
 
            renderMyEmbed(element, renderData);
        });
    }
}

export async function renderMyEmbed(element: Element, data: IEmbedData) {
    // Load up an external script that our embed depends on.
    await ensureScript("//platform.my-embed-site.com/widgets.js");
 
	// Ensure that the library we just loaded is available
	// (There could have been an error, network timeout, etc.)
    if (!window.globalEmbedLibrary) {
        throw new Error("MyEmbedSite library failed to load.");
    }
    // ... Do some rendering into `element` here.
}
```

**Pull request merged**